doc/designs/quic-design/quic-ackm.md

```diff
@@ -0,0 +1,411 @@
+#define QUIC_PN_SPACE_NUM           3
+```
+
+Packet numbers are 62-bit values represented herein by `QUIC_PN`.
+`QUIC_PN_INFINITE` evaluates to an invalid QUIC packet number value.
+
+```c
+/* QUIC packet number representation. */
+typedef uint64_t QUIC_PN;
+#define QUIC_PN_INFINITE            UINT64_MAX
+```
+
+Instantiation
+-------------
+
+The QUIC ACK manager is instantiated as follows:
+
+```c
+typedef struct quic_ackm_st QUIC_ACKM;
+
+QUIC_ACKM *QUIC_ACKM_new(OSSL_TIME (*now)(void *arg),
+                         void *now_arg,
+                         QUIC_RTT *rtt,
+                         OSSL_CC_METHOD *cc_method,
+                         OSSL_CC_DATA *cc_data);
+void QUIC_ACKM_delete(QUIC_ACKM *ackm);
+```
+
+The function pointer `now` is invoked by the ACK manager to obtain the current
+time. `now_arg` is passed as the argument. The congestion controller method and
+instance passed are used by the ACK manager instance. `rtt` points to an RTT
+tracker instance (discussed below).
+
+Events
+------
+
+The ACK manager state is evolved in response to events provided to the ACK
+manager by the user.
+
+### On TX Packet
+
+This must be called when a packet is transmitted. It does not provide the
+payload of the packet, but provides metadata about the packet which is relevant
+to the loss detection and acknowledgement process.
+
+The caller is responsible for the allocation of the structure and the structure
+must remain allocated until one of the callbacks is called or the ACK manager is
+freed. It is expected this structure will usually be freed (or returned to a
+pool) in the implementation of either callback passed by the user.
+
+Only exactly one of the callbacks in the structure will be called over the
+lifetime of a `QUIC_ACKM_TX_PKT`, and only once.
+
+Returns 1 on success.
+
+```c
+typedef struct  quic_ackm_tx_pkt_st {
+    /* The packet number of the transmitted packet. */
+    QUIC_PN pkt_num;
+    /* THe number of bytes in the packet which was sent. */
+    size_t num_bytes;
+    /* The time at which the packet was sent. */
+    OSSL_TIME time;
+    /*
+     * One of the QUIC_PN_SPACE_* values. This qualifies the pkt_num field
+     * into a packet number space.
+     */
+    unsigned int pkt_space :2;
+    /* 1 if the packet is in flight. */
+    unsigned int is_inflight :1;
+    /* 1 if the packet has one or more ACK-eliciting frames. */
+    unsigned int is_ack_eliciting :1;
+    /* 1 if the packet is a PTO probe. */
+    unsigned int is_pto_probe :1;
+    /* 1 if the packet is an MTU probe. */
+    unsigned int is_mtu_probe :1;
+
+    /* Callback called if frames in this packet are lost. arg is cb_arg. */
+    void (*on_lost)(void *arg);
+    /* Callback called if frames in this packet are acked. arg is cb_arg. */
+    void (*on_acked)(void *arg);
+    /* 
+     * Callback called if frames in this packet are neither acked nor lost. arg
+     * is cb_arg.
+     */
+    void (*on_discarded)(void *arg);
+    void  *cb_arg;
```

Comment:
We will want to put a type to this (once it's type is known).  I'm currently using `OSSL_QUIC_PACKET` but it will probably change.

Reply:
I feel like this is unnecessary and just couples things unnecessary.

Reply:
We've ongoing pain where we overused void pointers in the provider APIs :(